### PR TITLE
Event buffer

### PR DESCRIPTION
**Pull request by mdpye**
These changes take advantage of protocol version 8 and event-buffer to
catch up on messages missed while offline.

### CODE REVIEW

**Comment by zimbatm** [comment]
LGTM. At some point we will need a `subscribe(resumeAfter)` option but this is good for people who just want to upgrade their existing code and get the benefits of EB.

**Comment by mdpye** [comment]
FAO @zimbatm @leggetter

File: src/main/java/com/pusher/client/channel/ChannelEventListener.java [hunk 1]
```diff
@@ -39,6 +39,28 @@
...
+     * @param historyComplete
+     *            If true, the requested history was available in full.
+     *            If false, the requested history was not fully available, but
+     *            the subcription has been made. available history will be delivered
+     *            to the appropriate {@link SubscriptionEventListener#onEvent}
+     *            callbacks.
+     *            If null, history was not requested for this subscription.
      */
-    void onSubscriptionSucceeded(String channelName);
+    void onSubscriptionSucceeded(String channelName, Boolean historyComplete);
```
Comment (leggetter) [inline src/main/java/com/pusher/client/channel/ChannelEventListener.java]: We've generally avoided the term "history" so maybe we should also do that here.

That aside, `resumeSuccessful` may be more relevant/symmetrical to `resumeAfterId`.

File: src/main/java/com/pusher/client/channel/impl/ChannelImpl.java [hunk 2]
```diff
@@ -80,18 +83,56 @@ public void unbind(final String eventName, final SubscriptionEventListener liste
...
+        final String data = (String)jsonObject.get("data");
+        final String eventId = (String)jsonObject.get("id");
+
+        final Map<String, Object> dataMap = new Gson().fromJson(data, Map.class);
 
         if (event.equals(SUBSCRIPTION_SUCCESS_EVENT)) {
             updateState(ChannelState.SUBSCRIBED);
+
+            final Map<String, Object> resumeData = (Map<String, Object>)dataMap.get("resume");
+            final Boolean historyAvailable = resumeData == null ? null : (Boolean)resumeData.get("ok");
```
Comment (leggetter) [inline src/main/java/com/pusher/client/channel/impl/ChannelImpl.java]: See previous comment RE: `history`

**Pull request closed without merging**